### DISCUSSION TITLE
TShock Tutorial

### DISCUSSION THREAD

**Discussion by jcpt928** [Pryaxis/TShock#2384] [Q&A]
I really shouldn't have to ask this, after having run TShock for multiple Terraria servers for at least the last several years; but, there must have been one hell of a change with the way things work since I last had to rebuild a server.

All the existing TShock tutorials and documentation I've gone through seem to be outdated at best (both first-party and third-party), as command line options like "worldpath", "world", "config", either do not work, or explicitly return a "command not recognized" response in-script and\or at the command line. I notice now Terraria is forcing world saves into the <user>/.local folder as well, which, for servers, is just plain moronic. Yes, I know the above commands are supposed to allow you to change that; but, nothing I've tried seems to work in that regard.

Additionally, TShock doesn't load the config.json file it creates by default? What's the point of that file then if you subsequently have to force it to load in-script, and, right now, the documented way of doing so does not work? The Terraria serverconfig.txt file also does not seem to be recognized by the base server executable, and the commands for forcing it to do so don't seem to be working either.

Perhaps I'm missing something really big; but, it used to be as simple as creating a folder, downloading and unzipping TShock in that folder, running the server the first time and doing the setup, and then shutting it down to reconfigure everything through the config.json file and add pre-built worlds into the start-up (the ENTIRE server should run out of a single folder tree, and care less where it is on the system - partially so you can run multiple servers on a system). I've not even gotten back to the point where I can reattach to a MySQL database; so, any thoughts or pointers would be appreciated.

BTW, I have noticed some rather useless "troll" responses on other recent discussion threads - to people like that, get off GitHub.

**Comment by bartico6** [comment]
> All the existing TShock tutorials and documentation I've gone through seem to be outdated at best (both first-party and third-party), as command line options like "worldpath", "world", "config", either do not work, or explicitly return a "command not recognized" response in-script and\or at the command line. I notice now Terraria is forcing world saves into the /.local folder as well, which, for servers, is just plain moronic. Yes, I know the above commands are supposed to allow you to change that; but, nothing I've tried seems to work in that regard.

Flags are input parameters passed on the command line: `TerrariaServer.exe -flag value -flag2 value`. Commands are typed into the server console. Don't mix them up, since the server will, correctly, not recognize commandline flags as server commands (or worse so, could misinterpret them as a command running with unlimited (console) privileges.

The flags you're looking for are `-config` (for the "Terraria Server config" from the core game), `-configpath` for where tshock configs are supposed to reside, `-worldselectpath` for where the game will look for worlds in the startup listing, `-world` for the path to where the world to be auto-loaded is.

>  I notice now Terraria is forcing world saves into the /.local folder as well, which, for servers, is just plain moronic.

This is not our fault. The Terraria game server, upon startup, will look for a `-savedirectory` folder which it'll use to place the 'Terraria' folder with server data inside, or it'll default to whatever their internal system-specific utility class tells them. Which, on Linux, appears to default to ~/.local/share, as evidenced by them looking up the `HOME` env variable and appending .local/share to it.
![Decompiled source code of the class used for path determination](https://i.quake.tf/1625067629-42138749.png)

> The Terraria serverconfig.txt file also does not seem to be recognized by the base server executable, and the commands for forcing it to do so don't seem to be working either.

First, the game will only load that file if you tell it to using the `-config` option. Second, provided the option is set, the game code doesn't hint at any reason why it wouldn't load - the game takes the argument passed to `-config` and opens the file and reads it line by line:
![methodcall into loaddedconfig](https://i.quake.tf/1625067841-13195641.png)
![loaddedconfig itself](https://i.quake.tf/1625067873-13963820.png)

**Finally, if you are able to determine that some of these options dont work even though you used them correctly, open an [issue!](https://github.com/Pryaxis/TShock/issues/new)**

**Reply by jcpt928** [reply]
I appreciate the quick response - definitely not confusing flags and server console commands.

I agree that the default save folder is not TShock's fault - just had to point out how moronic it was regardless. :P

I will go ahead and open up issues for those I'm seeing not working. I kind of get the feeling that Terraria has actually made things slightly more difficult for you guys considering this doesn't seem to work quite as simple as it used to. I'd hate to have to revert to using Windows for this - try to run certain servers I've got on Linux for good reason.

**Comment by hakusaro** [comment]
> Additionally, TShock doesn't load the config.json file it creates by default? What's the point of that file then if you subsequently have to force it to load in-script, and, right now, the documented way of doing so does not work?

what?
